### PR TITLE
FileDropExtension

### PR DESCRIPTION
**Pull request by brapifra**
Hey!
First of all, congratulations on this awesome project :tada:
I think it really suits my needs for a personal project I have, so I will be adding some extensions in the near future.

This is the first one, which is quite simple, but I'm open to suggestions!

Have you thought about creating some docs for any new module/extension?

### CODE REVIEW

File: src/extensions/FileDropExtension/FileDropExtension.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+      this.loadVideo(videoFile);
+    }
+  }
+
+  private loadVideo = (videoFile: File) => {
+    const player = this.instance.player as HTML5Player;
+
+    if (!player.mediaElement.paused) {
+      this.instance.once(Events.STATE_DURATION_CHANGE, () => {
+        this.instance.play();
```
Comment (brapifra) [inline src/extensions/FileDropExtension/FileDropExtension.ts]: I don't like this `once` statement. Is there a better way of checking when the player has loaded the new file set by `player.setSource`?

File: src/extensions/FileDropExtension/FileDropExtensionLoader.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import { FileDropExtension } from '@src/extensions/FileDropExtension/FileDropExtension';
+import { Instance } from '@src/Instance';
+import { Config, IModuleLoader, ModuleLoaderTypes } from '@src/types';
+
+export const FileDropExtensionLoader = {
+  type: ModuleLoaderTypes.EXTENSION,
+
+  create: async (instance: Instance) => new FileDropExtension(instance),
+
+  isSupported: ({ config }: { config: Config }): boolean => !!config.fileDrop,
```
Comment (brapifra) [inline src/extensions/FileDropExtension/FileDropExtensionLoader.ts]: Also, I think there should be a way of accessing the instance here to do something like:

```typescript
isSupported: ({ config }: { config: Config }): boolean => !!config.fileDrop && instance.player instanceof HTML5Player
```

File: src/extensions/CaptionsExtension/CaptionsExtension.ts [hunk 3]
```diff
@@ -1,40 +1,40 @@
...
 
 export class CaptionsExtension extends Module {
   public name: string = 'CaptionsExtension';
 
-  private tracks: HTMLTrackElement[];
-
   constructor(instance: Instance) {
     super(instance);
 
-    this.tracks = instance.config.captions.map(caption => {
```
Comment (brapifra) [inline src/extensions/CaptionsExtension/CaptionsExtension.ts]: It's kind of useless to save the tracks when they can be gotten from `mediaElement.textTracks`, don't you think?

**Comment by matvp91** [comment]
Hey!

Thank you for your contribution, highly appreciated! There are multiple changes here that I'd like to address separately:

1) FileDropExtension

However, I'm a bit hesitant about including any possible extension to the core library. I'd like to restrict the default extensions to absolute must-haves for a video player in general (eg: ad logic / fullscreen and state management).

At the moment I am working on a spec that should allow others to write custom extensions / media / controllers and players by using the `IndigoPlayer.addModuleLoader(module)` API. This means that you can write an extension and host it in it's own GitHub repository (with proper versioning & once published on NPM, it's available through jsdelivr). Once finished, I'll provide a repository with an example extension using the addModuleLoader API. Would you mind waiting a few days so I can finish this? Afterwards, I can assist you in setting up your extension as an external one.

I'd be more than happy to include a link to your custom extension in the documentation.

2) CaptionsExtension

Would you mind isolating this change in a separate pull request? I'd be happy to merge it in for you.

Cheers!

**Comment by brapifra** [comment]
Hi again,

**FileDropExtension**

Yup, I totally agree with you about separating the core lib from the extensions. I'm looking forward to trying this new `addModuleLoader` feature.

Also, are you planning to include a way of adding UI items dynamically? It could be useful to be able to add some buttons within the extension class.

**CaptionsExtension**
Will do. I'm planning to add a button to select the captions too. Are you ok with that?

Cheers

**Comment by matvp91** [comment]
I've opened a ticket related to the `addModuleLoader` API: https://github.com/matvp91/indigo-player/issues/15, I'll keep progress visible there.

Regarding your questions: For now, I won't include a way for others to add UI items dynamically, the default UI will be shipped with core functionality only. People can easily build a custom UI and use the default UiExtension as a base reference (or developers are free to build their own UI from scratch). The reason for a basic UI is to offer a package that just works out of the box with the very basis of what a player UI should have.

As for the captions, I plan on introducing a `Settings` component in the near future, this will contain features related to the core such as bitrate (quality) selection, the selection of subtitles. I'll make sure I have that component shipped before I start the implementation of a subtitles selectbox so you can have a play with it already.

**Pull request closed without merging**